scripts/nn/examples/AttentionExample.dml

```diff
@@ -0,0 +1,442 @@
+      [W_2, mW_2, vW_2] = adam::update(W_2, dW_2, lr, beta2, beta2, epsilon, t, mW_2, vW_2)
+
+      # put weights & biases into list
+      biases = list(b_0,b_1,b_2)
+      weights = list(W_0,W_1,W_2,W_E)
+
+    }
+    val_loss = evaluate(x_val, y_val, biases, weights, lstm_neurons, max_sequence_length, embedding_size, batch_size)
+    loss = loss / iters
+    print("Epoch: " + ep + "; Train Loss: " + loss + "; Val. Loss: " + val_loss)
+  }
+}
+
+predict = function( matrix[double] x,
+                    List[unknown] biases,
+                    List[unknown] weights,
+                    integer max_sequence_length,
+                    integer embedding_size,
+                    integer lstm_neurons)
+  return (matrix[double] y_hat, matrix[double] out5, matrix[double] out4, matrix[double] out3,
+          matrix[double] out2, matrix[double] out1, matrix[double] emb, matrix[double] cache_out_out,
+          matrix[double] cache_c_out, matrix[double] cache_ifog_out)
+{
+  /*
+   * Predicts an output y_hat for given samples x.
+   *
+   * Inputs:
+   * - x: sample features of shape(batch_size, max_sequence_length)
+   * - biases: list of biases of length 3 (lstm, affine, affine)
+   * - weights: list of weights of length 4 (lstm, affine, affine, embedding)
+   * - max_sequence_length: number of words per sample.
+   * - embedding_size: size of embedding vector for 1 word
+   * - lstm_neurons: number of neurons in lstm layer.
+   *
+   * Outputs:
+   * - y_hat: matrix of shape(batch_size, 1), prediction for log-loss, output of sigmoid layer
+   * - out5: output of final affine layer, shape(batch_size, 1)
+   * - out4: output of relu layer
+   * - out3: output of hidden affine layer
+   * - out2: output of attention layer
+   * - out1: output states from lstm layer, of shape(batch_size, max_sequence_lengt * lstm_neurons)
+   * - cache_out_out: cache_out output from lstm layer
+   * - cahce_c_out: cache_c output from lstm layer
+   * - cache_ifog_out: cahce_ifog output from lstm layer
+   */
+
+  # unpack weights & biases
+  W_0 = as.matrix(weights[1])
+  W_1 = as.matrix(weights[2])
+  W_2 = as.matrix(weights[3])
+  W_E = as.matrix(weights[4])
+
+  b_0 = as.matrix(biases[1])
+  b_1 = as.matrix(biases[2])
+  b_2 = as.matrix(biases[3])
+
+  # fetch embedding
+  emb = fetch_embeddings(x, W_E, max_sequence_length, embedding_size)
+  # put input through layers
+  batch_size = nrow(x)
+  out0 = matrix(0, batch_size, lstm_neurons)
+  c0 = out0
+  [out1, c_out, cache_out_out, cache_c_out, cache_ifog_out]=
+    lstm::forward(emb, W_0, b_0, max_sequence_length, embedding_size, TRUE, out0, c0)
+  out2 = attention::forward(out1, out1, out1, max_sequence_length)
+  out3 = affine::forward(out2, W_1, b_1)
+  out4 = relu::forward(out3)
+  out5 = affine::forward(out4, W_2, b_2)
+  y_hat = sigmoid::forward(out5)
+}
+
+fetch_embeddings = function(matrix[double] indexes, matrix[double] W_E,
```

Comment:
not sure either, but we could limit the example to not include embedding, to make the project simpler :) since it goes out of scope of attention models anyway.